csi/src/findmnt.rs

```diff
@@ -0,0 +1,223 @@
+use crate::error::DeviceError;
+use serde_json::Value;
+use std::{collections::HashMap, process::Command, string::String, vec::Vec};
+
+// Keys of interest we expect to find in the JSON output generated
+/// by findmnt.
+const TARGET_KEY: &str = "target";
+const SOURCE_KEY: &str = "source";
+const FSTYPE_KEY: &str = "fstype";
+
+#[derive(Debug)]
+pub struct DeviceMount {
+    pub mount_path: String,
+    pub fstype: String,
+}
+
+#[derive(Debug)]
+struct FindmntFilter<'a> {
+    key: &'a str,
+    value: &'a str,
+}
+
+impl PartialEq<Value> for FindmntFilter<'_> {
+    /// Special case the comparison for the source field returned
+    /// by findmnt.
+    fn eq(&self, value: &Value) -> bool {
+        if self.key == SOURCE_KEY {
+            if let Some(strvalue) = value.as_str() {
+                let devpath = convert_findmnt_devicepath(strvalue);
+                if devpath == self.value {
+                    return true;
+                }
+            } else {
+                error!("value for {} is not a string", self.key);
+            }
+        }
+        self.value == value
+    }
+}
+
+/// The source field returned from findmnt and
+///   can be different for the same source on different systems,
+///   for example
+///   dev[/nvme0n1], udev[/nvme0n1], tmpfs[/nvme0n1], devtmpfs[/nvme0n1]
+///   Convert this to the expected /dev/nvme0n1 and added to the hashmap
+fn convert_findmnt_devicepath(devpath: &str) -> String {
+    lazy_static! {
+        static ref RE_UDEVPATH: regex::Regex = regex::Regex::new(
+            r"(?x).*\[(?P<device>/.*)\]
+        ",
+        )
+        .unwrap();
+    }
+    match RE_UDEVPATH.captures(devpath) {
+        Some(caps) => format!("/dev{}", &caps["device"]),
+        _ => devpath.to_string(),
+    }
+}
+
+/// Convert the json map entry to a hashmap of strings
+/// The source field returned from findmnt is converted
+/// to the /dev/xxx form if required.
+fn jsonmap_to_hashmap(
```

Comment:
Again TryFrom() is more appropriate here. Also, I'm not sure why we need to convert here? It already is a Map<String, Value>

Reply:
I replaced the use of json map clone because  the callers always want strings, and not having to deal with Value makes those bits of code simpler to read.

Reply:
Cannot do TryFrom since neither struct is defined in findmnt. I have simplified the function.